lib/vrl/stdlib/src/to_regex.rs

```diff
@@ -0,0 +1,70 @@
+use regex::Regex;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ToRegex;
+
+impl Function for ToRegex {
+    fn identifier(&self) -> &'static str {
+        "to_regex"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::BYTES,
+            required: true,
+        }]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "regex",
+                source: "to_regex(s'^foobar$')",
+                result: Ok("r'^foobar$'"),
+            }
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        Ok(Box::new(ToRegexFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ToRegexFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ToRegexFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?.try_bytes()?;
+        let regex = regex::Regex::new(value).map_err(|err| format!("could not create regex: {}",err))?.map(Into::into);
```

Comment:
Hi @jszwedko, I'm getting some compiler errors here, not sure how to fix them. I think the problem is that we should be coercing the value into a `regex` type for VRL instead of a `regex::Regex` type in Rust. 
```
error[E0308]: mismatched types
  --> lib/vrl/stdlib/src/to_regex.rs:44:39
   |
44 |         let regex = regex::Regex::new(value).map_err(|err| format!("could not create regex: {}",err))?.map(Into::into);
   |                                       ^^^^^ expected `&str`, found struct `bytes::Bytes`

error[E0599]: the method `map` exists for struct `regex::Regex`, but its trait bounds were not satisfied
   --> lib/vrl/stdlib/src/to_regex.rs:44:104
    |
44  |         let regex = regex::Regex::new(value).map_err(|err| format!("could not create regex: {}",err))?.map(Into::into);
    |                                                                                                        ^^^ method cannot be called on `regex::Regex` due to unsatisfied trait bounds
```

Reply:
The type of `value` here is `bytes`. Regex wants it the be `&str`. To convert you can use [`str::from_utf8`](https://doc.rust-lang.org/std/str/fn.from_utf8.html). Something like this should do the trick:

```rust
  let value = str::from_utf8(value).map_err(|err| format!("regex was not a valid utf8 string {}", err))?;
```

[Regex::new](https://docs.rs/regex/1.4.5/regex/struct.Regex.html#method.new) returns a `Result<Regex, Error>`. By appending the call with a `?` it makes the whole function return an error should that Result be `Err`, otherwise it will resolve to the Regex. So here it is calling `map(Into::into)` on the `Regex`, but `Regex` doesn't have a `map` function defined against it. The error message says it does, but I can't find anywhere in the documentation for it, so it is likely to be some trait somewhere that sort of does, but doesn't - hence the error message. 

You could either move the `?`, so you would then be calling `map` against the `Result` type:

```rust
 let regex = regex::Regex::new(value).map_err(|err| format!("could not create regex: {}",err)).map(Into::into)?;
```

Or you could keep the `?` where it is and just call `into` against the regex directly to convert it to a vrl value:

```rust
 let regex = regex::Regex::new(value).map_err(|err| format!("could not create regex: {}",err))?.into();
```

If it complains that it doesn't know what type it should convert into, specify it directly:

```rust
let regex: Value = ...
```